### ISSUE TITLE
Issue with KISS prediction [Bug]

### ISSUE THREAD

**Issue by g-benton** [cornellius-gp/gpytorch#836]
# 🐛 Bug

When producing posterior predictions using `GridInterpolationKernel` in two dimensions I am seeing an error where the predictions are constant along one of the axes around the edge of the predicted surface (see image below). The test points are on a grid that is contained within the bounds of the inducing grid so I think the kernel interpolation should be possible. 

Just a note: I'm setting up the data so that the inducing points in the interpolation kernel are the same as the training points, but that is intentional.

I've been taking apart the `GridInterpolationKernel` code to generate posterior predictions for a special use case and am seeing the same artifacts in posterior predictions. So if this is in fact a bug and not an error with how I'm setting up data I would guess it is in `_compute_grid` or `InterpolatedLazyTensor` as those are the only components I'm using. 

<img width="273" alt="bug" src="https://user-images.githubusercontent.com/29232979/62789277-6fbf5800-ba7d-11e9-980a-ddaa19660e3a.png">

## To reproduce
```python
## imports and data 
import math
import torch
import gpytorch
from gpytorch.kernels import GridInterpolationKernel, PeriodicKernel, ProductKernel
import matplotlib.pyplot as plt

grid_size = 10
grid_bounds = [(0., 10.), (0., 10.)]

train_grid = torch.zeros(grid_size, len(grid_bounds))
for i in range(len(grid_bounds)):
    grid_diff = float(grid_bounds[i][1] - grid_bounds[i][0]) / (grid_size - 2)
    train_grid[:, i] = torch.linspace(grid_bounds[i][0] - grid_diff, grid_bounds[i][1] + grid_diff, grid_size)

train_x = gpytorch.utils.grid.create_data_from_grid(train_grid)
train_y = torch.sin(train_x[:, 0]) + torch.sin(train_x[:, 1])

## model set up
class GPRegressionModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood,
                grid_size, grid_bounds):
        super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)

        self.mean_module = gpytorch.means.ConstantMean()

        self.covar_module = GridInterpolationKernel(
                gpytorch.kernels.RBFKernel(ard_num_dims=2),
                grid_size=grid_size, grid_bounds=grid_bounds)

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = GPRegressionModel(train_x, train_y, likelihood,
                         grid_size, grid_bounds)

## You can train but it's not really important to the issue
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam([
    {'params': model.parameters()},  # Includes GaussianLikelihood parameters
], lr=0.1)

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

def train():
    training_iterations = 100
    for i in range(training_iterations):
        optimizer.zero_grad()
        output = model(train_x)
        loss = -mll(output, train_y)
        loss.backward()
        print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
        optimizer.step()

train()

## testing data and predictions
test_grid_sz = 500
test_grid_bds = [(0, 10), (0, 10)]

test_grid = torch.zeros(test_grid_sz, len(test_grid_bds))
for i in range(len(test_grid_bds)):
    grid_diff = float(test_grid_bds[i][1] - test_grid_bds[i][0]) / (test_grid_sz - 2)
    test_grid[:, i] = torch.linspace(test_grid_bds[i][0] - grid_diff, test_grid_bds[i][1] + grid_diff, test_grid_sz)

test_x = gpytorch.utils.grid.create_data_from_grid(test_grid)

model.eval()
with gpytorch.settings.skip_posterior_variances(True):
    pred_y = model(test_x).mean

plt.imshow(pred_y.view(test_grid_sz, test_grid_sz).detach().numpy())
plt.title("Predicted Surface")
```

## Expected Behavior

I would expect the prediction surface to be smooth in all directions out to the edges. 

## System information

**Please complete the following information:**
- GPyTorch Version 0.3.2

- Pytorch Version 1.1.0
- Computer OS: macOS 10.14.4

**Comment by jacobrgardner** [comment]
@g-benton Yeah, so you are running in to boundary issues -- the cubic interpolation algorithm does strange things when you have data between `(grid_lower_bound, grid_lower_bound + 2 * grid_pt_width)`. 

In GPyTorch, we solve this problem internally by taking the grid bounds you give us and extending it on the left and the right by two grid widths so that no data actually lies on the boundary. However, by constructing the grids in a custom fashion using the same code, you've kind of circumvented our sneaky fix by placing training data out of the grid bounds.

Here's a few potential fixes:

1. For the GPyTorch kernel, don't bother specifying the bounds because we can determine them from the data. If you just create your kernel as 
```python
self.covar_module = GridInterpolationKernel(
                gpytorch.kernels.RBFKernel(ard_num_dims=2),
                grid_size=grid_size, num_dims=2)
```
Then you get:
![image](https://user-images.githubusercontent.com/4016393/62801055-b28f2900-ba99-11e9-97ff-cbbd4b467890.png)

2. Since both your training data and test data actually lie on an exact grid, consider using `GridKernel` directly instead of `GridInterpolationKernel`.

**Comment by g-benton** [comment]
Thank you for the speedy reply. That was what I had suspected (but feared). What is confusing about this is that I don't see these artifacts when I move the testing points away from the boundaries (nor do I see them along all edges of the predicted surface) - which makes it seem like setting up the training data like this is fine and that the issue is with the testing points but perhaps I'm missing something here?

For my particular use case I want to assume the training data is on a grid and use the training points as the inducing points. I can compute `K(X, X)` exactly, but with the kernel I am using `K(X, X*)` is very expensive to compute - so I'm only calling the interpolation routine to make posterior predictions.

**Comment by jacobrgardner** [comment]
I think the right way to do what you are doing is then a hybrid of the two kernels -- you want `K_{XX}` and possibly `K_{X*X*}` to be a `ToeplitzLazyTensor` rather than an `InterpolatedLazyTensor`, but you want interpolation when calling with `X*` and `X`.

Maybe just write a quick `Kernel` whose forward calls a `GridKernel` when the data it is passed (`x1, x2`) are equal and calls a `GridInterpolationKernel` when they aren't?

**Comment by g-benton** [comment]
That's similar to what I'm doing. I'm really only interested in posterior means, so I've just written a custom routine to compute `W* K(X, X) W K(X, X)^{-1} y \approx K(X*, X)K(X, X)^{-1} y`, which is working besides the issue from above. 

I'm still a bit confused that even when my testing set is completely contained in the inducing grid I'm getting artifacts in predictions.

**Comment by g-benton** [comment]
Hi @jacobrgardner - I'm wondering if you've given this any more thought. I'm continuing to see these artifacts in many cases when the testing points are well within the bounds of the inducing grid. 

Another thought that occurred to me is that I'm I have only ever the behavior shown above where the artifacts are only along two edges of the testing surface even when all bounds of the testing points are equidistant from the edges of the inducing grid. If this truly were an issue with how the testing points were being set up why would I not see these artifacts along all edges?

Sorry if these questions are silly - this behavior just feels very unexpected to me.

**Comment by jacobrgardner** [comment]
Hi @g-benton When you say well within the bounds of the grid, does this mean specifically by at least two grid widths? As in, you are specifically far enough within the grid that you should not be seeing the boundary condition but are?

I'm not sure why that boundary effect is showing up only on two edges -- I can look in to that.

Could you share some specifics on the custom routine you wrote for predictive means since this might be helpful? Are you calling `GridInterpolationKernel` between the test and train data and `GridKernel` to get `K(X, X)`? I'm wondering if something is happening here specifically, because at least when I test things where the test points truly lie well within the grid bounds, I do not see problems at the boundary.
